uspace/lib/cpp/src/__bits/test/future.cpp

```diff
@@ -0,0 +1,351 @@
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <__bits/test/mock.hpp>
+#include <__bits/test/tests.hpp>
+#include <chrono>
+#include <exception>
+#include <future>
+#include <tuple>
+#include <utility>
+
+using namespace std::chrono_literals;
+
+namespace
+{
+    template<class R>
+    auto prepare()
+    {
+        auto res = std::tuple<
+            std::promise<R>, std::future<R>,
+            std::aux::shared_state<R>*
+        >{};
+        std::get<0>(res) = std::promise<R>{};
+        std::get<1>(res) = std::get<0>(res).get_future();
+        std::get<2>(res) = std::get<1>(res).__state();
+
+        return res;
+    }
+}
+
+namespace std::test
+{
+    bool future_test::run(bool report)
+    {
+        report_ = report;
+        start();
+
+        test_future();
+        test_promise();
+        test_future_promise();
+        test_async();
+        test_packaged_task();
+        test_shared_future();
+
+        return end();
+    }
+
+    const char* future_test::name()
+    {
+        return "future";
+    }
+
+    void future_test::test_future()
+    {
+        std::future<int> f1{};
+        test("default constructed invalid", !f1.valid());
+
+        std::future<int> f2{new std::aux::shared_state<int>{}};
+        test("state constructed valid", f2.valid());
+
+        f1 = std::move(f2);
+        test("move assignment source invalid", !f2.valid());
+        test("move assignment destination valid", f1.valid());
+
+        std::future<int> f3{std::move(f1)};
+        test("move construction source invalid", !f1.valid());
+        test("move construction destination valid", f3.valid());
+    }
+
+    void future_test::test_promise()
+    {
+        std::promise<int> p1{};
+        test("default constructed promise has state", p1.__state());
+
+        std::promise<int> p2{};
+        auto* s1 = p1.__state();
+        auto* s2 = p2.__state();
+        p2.swap(p1);
+        std::swap(s1, s2);
+
+        test_eq("swap switches states pt1", s1, p1.__state());
+        test_eq("swap switches states pt2", s2, p2.__state());
+
+        std::promise<int> p3{std::move(p1)};
+        test_eq("move construction state moved", s1, p3.__state());
+        test_eq("move construction source empty", p1.__state(), nullptr);
+
+        p1 = std::move(p3);
+        test_eq("move assignment state move", s1, p1.__state());
+        test_eq("move assignment source empty", p3.__state(), nullptr);
+
+        p1.set_value(42);
+        test("set_value marks state as ready", s1->is_set());
```

Comment:
Now that's a tricky one. I can add a check for that if you wish, but the reason for not checking it was that we don't have exception support yet. This is because that if one calls `set_value` when the shared state is already marked as ready, an exception with the error code `std::future_errc::promise_already_satisfied` is thrown, but right now throwing an exception results to `abort` being called (and a message being printed to stderr). Because of the abort it was not feasible to test for it, but if you want I can add a check for `!is_set()` on the shared state before `set_value` is called (though it is IMHO redundant because of the exception/abort).

Reply:
I think there are more places where you check the final value, but don't check the initial value, and so you don't actually know whether calling a function results in the desired change. From your reply I understand that calling is_set() before set_value() is okay provided the tested code is correct? Then I'd do it. Unfortunately, I didn't track all the other occurrences that follow the same pattern.

Reply:
> From your reply I understand that calling is_set() before set_value() is okay provided the tested code is correct?

The `is_set` member function is purely internal because it is implemented on the shared state (which itself is an implementation detail), so there should be no problem testing it before. I will check when I'm home just to be sure, though I might just merge this and add this kind of testing to my TODO list in my cpp-qol [0] branch which I'm working on.

[0] https://github.com/Dzejrou/helenos/tree/cpp-qol